ZSA test vectors modifications

PaulLaux opened the pull request:
Updated reference implementation for the TXV6 and OrchardZSA components.

daniben31 committed:
implement text vectors isk and ik

daniben31 committed:
generate test vectors for issuance keys

daniben31 committed:
Merge pull request #2 from QED-it/issuer_keys

Issuer keys test vectors implementation

naure committed:
zsa-note-type: test vectors for note commitment and encryption (#1)

* zsa-note-type: test vectors for note commitment and encryption

* zsa-note-type: explicit native then ZSA notes

* zsa-note-type: consistent order of arguments

* zsa-note-type: revert output format

* zsa-note-type: commit note encryption with note type test vectors

Co-authored-by: Aurélien Nicolas <info@nau.re>

naure committed:
ZSA update rust code (#3)

* zsa-sync-test-vectors: output directly the needed test code

* zsa-sync-test-vectors: regenerate rust code with the same content but final format

Co-authored-by: Aurélien Nicolas <info@nau.re>

alexeykoren committed:
Asset ID (#5)

PaulLaux committed:
Testing encryption v3 (#6)

- Updated to support encryption_v3 as given in QED-it/orchard#38 and QED-it/librustzcash#18.

- This PR breaks compatibility with OrchardDomainV2

vivek-arte committed:
Changes to derive Asset base from the Asset identifier (#7)

This adds the changes to the zcash test vectors to update the calculation of the Asset Base from the Asset Identifier via the Asset Digest, as detailed in ZIP 227.

ConstanceBeguier committed:
Update R constant for ZSA note commitment (#8)

We would like to have the same R constant for ZEC and ZSA note commitments.

vivek-arte committed:
Change the Issuance Authorization Signature scheme to the BIP340 Schnorr scheme, and rename `asset_id` to `asset_base` (#12)

This updates the ik derivation to be via the Schnorr scheme, and refactors asset_id.py to asset_base.py.

It also adds test vector generation code for issuance authorization signature, adds a check to ensure the `isk` is non zero, and updates the generated Rust files.

vivek-arte committed:
Merge branch 'zsa1' into upstream_change_merge

vivek-arte committed:
Upstream change merge (#15)

This merges upstream changes into the `zsa1` branch of our fork.

vivek-arte committed:
Separate OrchardZSA from Orchard (#13)

This separates the additions of the ZSA work into a separate file, creating separate test vector files for them. It also adds in upstream changes to the test vectors crate.

vivek-arte committed:
Making suggested changes (#16)

* changing documentation of the BIP 340 code to reflect the specific commit it has been taken from

* re-adding the main function to key_components so that orchard vectors can be generated independently

* reducing code duplication between orchard and orchard_zsa key_components

* cosmetic changes to reduce unnecessary divergences from upstream, addition of blank line at end of some files

* making the key component random initialization cleaner

* adding suggestions from review

vivek-arte committed:
Merge branch 'master' into zsa_review
This is to add upstream changes for ZIP 320 test vectors into the repository.

vivek-arte committed:
updating newly merged Rust test vectors to follow format

vivek-arte committed:
change output rendering to reduce git diff -- changes to rust files in next commit

vivek-arte committed:
updating files based on previous commit

vivek-arte committed:
reverting transaction.py edits, will include in subsequent PR

vivek-arte committed:
Adding upstream changes and reducing the file diff from upstream (#21)

This PR adds in the changes for ZIP 320 vectors from upstream. It also
edits the output file in order to reduce the diff between the generated
test vectors upstream.

vivek-arte committed:
Changes to TXID generation to account for ZSA additions (#22)

* updating zip_0244 functions to support generating both V5 and V6 digests, and adding NU6 constants to transaction.py

* adding V6 transaction class to transaction.py -- code reuse needs to be improved

* changes to transaction.py and zip_0244.py

* minor update

* initial attempt at inheritance, to be improved

* changing inheritance structure of TransactionV5 and TransactionV6, and moving TransactionV6 to transaction_zsa.py

* reverting zip_0244.py to only include V5 details

* separate txid vectors for zsa into separate files

* adding ZSA version of zip 244 vectors

* renaming to NU7 and so on

* still WIP, updating to work with librustzcash code

* updating get_random_unicode_bytes to give a slightly wider range of values

* changes to the transaction format serialization and the txid generation to make it spec compliant

* updating the generated test vectors

* some cleanup changes

* removing unichr option from get_random_unicode_bytes

* cleaner import

* simple changes based on review comments

* moving common transaction fields in a way that reduces code duplication

* reducing code duplication by creating an OrchardActionBase class which is inherited by OrchardActionDEscription and OrchardZSAActionDescription

* reducing code duplication in transaction, transaction_zsa, zip_0244 and orchard_zsa/digests files

* reducing code duplication inside the main function of zip_0244 and orchard_zsa/digests

* changing txn version back to V6

* renaming rho to nf_old in note_encryption files

vivek-arte committed:
Rename TransactionZSA to TransactionV6 (#23)

This renames the TransactionZSA class to TransactionV6 to better reflect its purpose as a V6 transaction format reference implementation

vivek-arte committed:
Updating test vectors for Action Groups (#24)

ConstanceBeguier committed:
Replace asset_desc by asset_desc_hash (#25)

For more details, see ZIP227

ConstanceBeguier committed:
Merge branch 'master' into merge_master_20250408

vivek-arte committed:
Move burn fields into the Action Group (#28)

* Moving the burn fields into the action group

* fixing small issue with switch to asset_desc_hash

* regenerating test vector files

ConstanceBeguier committed:
Merge branch 'zsa1' into merge_master_20250408

ConstanceBeguier committed:
Run rustfmt on generated rust files

ConstanceBeguier committed:
Merge branch 'master' into zsa1 (#30)

Main modifications
- Generate test vectors for ZIP230 adresses
- Generate test vectors for ZIP32
- Update regenerate script
- Update rust output by replacing vectors by slices to support no_std

vivek-arte committed:
Update encodings for `ik` and `issueAuthSig` (#31)

This is a companion PR to QED-it/orchard#182. It makes the changes to the specification made in zcash/zips#1042, namely the addition of a 0x00 byte to the start of the issuance validating key and the issuance authorization signature.

ConstanceBeguier committed:
Add versioned signatures (#32)

Add `SighashInfo` (version and associated data) to
- Orchard binding signatures,
- Orchard authorizing signatures, and
- issuance authorization signatures.

Update `auth_digest` by adding `SighashInfo`

vivek-arte committed:
Updates to digests and transaction format corresponding to recent changes to ZIPs (#36)

There was a rearranging of the transaction format and the digests that took place in zcash/zips#1086 and zcash/zips#1088.

This PR makes the changes to the test vectors to match them back to the spec.

---------

Co-authored-by: Constance Beguier <constance@qed-it.com>

ConstanceBeguier committed:
Fix orchard_auth_digest (#37)

Remove nested loop in orchard_auth_digest

ConstanceBeguier committed:
Add spend_auth_sigs_auth_digest into orchard_auth_digest (#38)

* Add orchard_zsa_spend_auth_sigs_auth_digest

* Update test vectors

ConstanceBeguier committed:
Add versioned signatures for Sapling transactions (#34)

Add `SighashInfo` (version and associated data) to
- Sapling binding signatures, and
- Sapling authorizing signatures

Update `sapling_auth_digest` by adding `SighashInfo` as defined in ZIP246

ConstanceBeguier committed:
Add versioned signatures for Transparent transactions (#35)

Add `SighashInfo` (version and associated data) to
- Transparent authorizing signatures.

Update `auth_digest` by adding `SighashInfo` as defined in ZIP246

Add `SighashInfo` for Sapling/Orchard binding signatures only when the signatures exist

oxarbitrage commented on zcash_test_vectors/orchard_zsa/commitments.py:
```diff
@@ -0,0 +1,69 @@
+        group_hash(D_blind + b"-r", b"") * r
+    )
+
+# ZIP-226 (https://github.com/zcash/zips/pull/628)
```
```suggestion
# ZIP-226 (https://zips.z.cash/zip-0226)
```

oxarbitrage commented on zcash_test_vectors/orchard_zsa/digests.py:
```diff
@@ -0,0 +1,156 @@
+    if len(tx.vActionGroupsOrchard) > 0:
+        for ag in tx.vActionGroupsOrchard:
+            digest.update(orchard_zsa_actions_compact_digest(ag))
+            # TODO remove memo digests once the new memo bundles are implemented (ZIP-231)
```
```suggestion
            # TODO: remove memo digests once the new memo bundles are implemented (ZIP-231)
```

vivek-arte replied:
The memo bundle update is independent work, and these comments are not exhaustively covering all the required changes. Removing this for now to avoid confusion.

oxarbitrage commented on zcash_test_vectors/orchard_zsa/asset_base.py:
```diff
@@ -0,0 +1,118 @@
+
+    random.seed(rand.u8())
+
+    # Update this to include code point ranges to be sampled
```
```suggestion
    # TODO: Update this to include code point ranges to be sampled
```

oxarbitrage commented on zcash_test_vectors/orchard_zsa/commitments.py:
```diff
@@ -0,0 +1,69 @@
+# Commitment schemes used in Orchard https://zips.z.cash/protocol/nu5.pdf#concretecommit
+from ..orchard.commitments import rcv_trapdoor, L_ORCHARD_BASE
+
+def value_commit(rcv: Scalar, v: Scalar, asset: Point):
```
Is this the same as https://github.com/QED-it/zcash-test-vectors/blob/zsa1/zcash_test_vectors/orchard/commitments.py#L18 ? If so, maybe we should just import.

oxarbitrage commented on zcash_test_vectors/orchard_zsa/digests.py:
Please add reference or short text to all of the definitions here. For example, there is no reference to https://zips.z.cash/zip-0246 in the file where all the personalizations used in the definitions are specified.
Also, please reduce the double space between the definitions where possible.

oxarbitrage commented on zcash_test_vectors/orchard_zsa/key_components.py:
```diff
@@ -0,0 +1,131 @@
+# The algorithm byte prefix for the encoding of the BIP340 Schnorr signature in ZIP227 is 0x00.
+ZSA_BIP340_SIG_SCHEME = b'\0'
+
+#This function provides the encoding of the issuance key, with the algorithm byte prefix.
```
```suggestion
# This function provides the encoding of the issuance key, with the algorithm byte prefix.
```

vivek-arte replied:
Done in QED-it#44

oxarbitrage commented on zcash_test_vectors/transaction.py:
```diff
@@ -539,7 +574,7 @@ def __bytes__(self):
                 ret += bytes(desc) # Excludes spendAuthSig
             ret += struct.pack('B', self.flagsOrchard)
-            ret += struct.pack('<Q', self.valueBalanceOrchard)
+            ret += struct.pack('<q', self.valueBalanceOrchard)
```
Can we justify this change ?

vivek-arte replied:
For more details, see https://github.com/QED-it/zcash-test-vectors/pull/44#discussion_r2685864158

oxarbitrage reviewed (commented):
I left several comments related to cleanup, readability, and adding documentation. Applying some of these suggestions  (and moving a few of the unrelated changes into a separate PR) would make the diff smaller and cleaner, which in turn should help us review the critical parts more carefully.

That said, these are all optional suggestions and reflect personal preference; I didn’t find any critical issues in this initial review.

If you prefer, I can also push the suggested changes directly to your branch.

vivek-arte committed:
adding init_ functions for the different bundles in the transaction

vivek-arte committed:
splitting to_bytes into components for header, transparent and sapling in TransactionBase

vivek-arte committed:
small renaming to match conventions

vivek-arte committed:
Merge branch 'master' into upstream_merge_prep_merge_with_master

vivek-arte committed:
Preparation changes for merging with upstream (#45)

This performs some code refactoring to make the upstream merge in the next step less complicated.

The refactoring is cosmetic in that it doesn't affect the generated test vectors.
It mainly splits out the initialization and byte serialization of the transaction (in TransactionBase) to differing functions for header, transparent and sapling parts of the transaction.
The Orchard parts are in the child classes (TransactionV5 and TransactionV6) and are not separated out into functions.

PaulLaux committed:
Merge branch 'zsa1' into upstream_merge_prep_merge_with_master

vivek-arte committed:
Merge with upstream (#46)

This merges in the updates from `upstream/master` to the target
branch.

vivek-arte commented on zcash_test_vectors/orchard/note.py:
```diff
@@ -2,7 +2,7 @@
 
 from .commitments import note_commit
 from .key_components import diversify_hash, prf_expand, derive_nullifier, FullViewingKey, SpendingKey
-from .pallas import Point, Scalar
```
Done in https://github.com/QED-it/zcash-test-vectors/pull/43 - if that work is merged, it will keep the diff of this PR smaller

vivek-arte commented on zcash_test_vectors/orchard_zsa/asset_base.py:
```diff
@@ -0,0 +1,118 @@
+def native_asset():
+    return group_hash(b"z.cash:Orchard-cv", b"v")
+
+
```
We use the two blank lines format in line with [PEP 8 style guide](https://peps.python.org/pep-0008/#blank-lines) for Python code.

vivek-arte commented on zcash_test_vectors/orchard_zsa/asset_base.py:
```diff
@@ -0,0 +1,118 @@
+    return h.digest()
+
+
+def encode_asset_id(key, asset_desc_hash):
```
Done in QED-it#44

vivek-arte commented on zcash_test_vectors/orchard_zsa/asset_base.py:
```diff
@@ -0,0 +1,118 @@
+    ]
+
+    alphabet = [
+        get_char(code_point) for current_range in include_ranges
```
Done in QED-it#44

vivek-arte commented on zcash_test_vectors/orchard_zsa/digests.py:
```diff
@@ -0,0 +1,156 @@
+    for desc in ag.vActionsOrchard:
+        digest.update(bytes(desc.cv))
+        digest.update(bytes(desc.rk))
+        # TODO remove encCiphertext[596:] once the new memo bundles are implemented (ZIP-231)
```
The memo bundle update is independent work, and these comments are not exhaustively covering all the required changes. Removing this for now to avoid confusion.

vivek-arte commented on zcash_test_vectors/orchard_zsa/asset_base.py:
```diff
@@ -0,0 +1,118 @@
+            ret.append(rng.randrange(0, 256))
+        return bytes(ret)
+
+    rand = Rand(randbytes)
```
I agree with it being better to import this from a single definition. However, I prefer to follow the convention of the repository as is (which is to duplicate). The follow-up cleanup PR can then perform a uniform sweep through all definitions to replace them with the imported definition. 

There may also be preferred conventions that I am unaware of about where the definition should be located, and it is better not to have these definitions point somewhere else.

vivek-arte commented on zcash_test_vectors/orchard_zsa/commitments.py:
```diff
@@ -0,0 +1,69 @@
+        while len(ret) < l:
+            ret.append(rng.randrange(0, 256))
+        return bytes(ret)
+    rand = Rand(randbytes)
```
See response in https://github.com/zcash/zcash-test-vectors/pull/108#discussion_r2684931362

vivek-arte committed:
Minor changes suggested in review (ZSA) (#44)

This makes the spacing, comment and minor rearrangements suggested in zcash#108 (review).
Specifically, this PR makes the changes that are touching the ZSA portions of the repository.

vivek-arte committed:
Refactoring the `SighashInfo` portion of the transaction (#49)

This refactors the portions of the `TransactionBase` class so that it does not need to use the `version_group_id` to decide behaviour - the way it behaves is decided by redefinition of the added functions in the child classes (aka `TransactionV5` and `TransactionV6`).

This refactor does not affect the generated test vectors.

vivek-arte committed:
revert rustfmt change

vivek-arte committed:
revert output visibility and indentation

vivek-arte committed:
regenerating test vectors

vivek-arte committed:
Revert formatting updates (#50)

This removes the formatting updates we made (namely the addition of
`rustfmt`, and the addition of `pub(crate)` visibility and indentation
changes).

This reduces the difference between `upstream:master` and our branch
significantly, from 69 files to 40 files, and would help speed up the
review of the ZSA test vectors.

vivek-arte committed:
Expand the sample set of Unicode code points (#47)

This expands the sampled code points for the get_random_unicode_bytes function to include all allowed Unicode code points.

This changes the random values of the description of the Assets, and therefore changes the orchard_zsa_digests.rs and orchard_zsa_asset_base.rs files, which would need to be copied into the librustzcash and orchard repositories respectively.

vivek-arte committed:
Add ZIP links to the definition of personalizations, and fix typo (#52)

This adds comments referencing the relevant sections of the ZIPs, as requested in zcash#108 (comment). We also update some constant names to fix a typo.

ConstanceBeguier committed:
Fix test vectors for ZIP233 (#53)

- Update nu7 branch id in ZIP233
- Add ZIP233 into `regenerate.sh` script
- Update ZIP233 test vectors

dmidem committed:
Generate canonical Orchard ZSA proof sizes in V6 test vectors

dmidem committed:
Trigger CI

dmidem committed:
Update json and zcash test vectors for zip_0233 and orchard_zsa_digests

PaulLaux committed:
Generate canonical proof size inside ActionGroupDescription

PaulLaux committed:
Drop stale transaction-level proofsOrchard on V6

PaulLaux committed:
Document source of Orchard ZSA proof-size constants

dmidem committed:
Move canonical proof sizing into ActionGroupDescription (#56)

- Generate the canonical-size proof inside
ActionGroupDescription.__init__ and drop the now-unused proofs_orchard
constructor parameter, so the size invariant is enforced by the class
rather than each call site.
- Delete the stale transaction-level proofsOrchard on V6 (the base class
draws it only for rand-stream stability), so accidental reads fail
loudly instead of using never-serialized bytes.

dmidem committed:
Generate canonical Orchard ZSA proof sizes in V6 test vectors (#55)

This PR updates the V6 transaction test-vector generator to produce
Orchard ZSA proofs with the canonical size expected by strict
deserialization.

The problem was encountered while syncing `QED-it/librustzcash` in
https://github.com/QED-it/librustzcash/pull/310. After strict Orchard
proof-size enforcement was introduced, the regenerated V6/ZSA vectors
caused the `zip_0233` and `zip_0244` transaction digest tests to fail in
`librustzcash`.

Previously, `TransactionV6` reused the random Orchard proof bytes
generated by the shared transaction base code. Those bytes could have
arbitrary short lengths, while Orchard ZSA requires the proof size to
match the number of actions in the action group.

This PR changes only the V6 generator (`transaction_v6.py`). It moves
Orchard ZSA proof-size generation into `ActionGroupDescription`, where
the number of Orchard actions is known, and generates the proof using:

```text id="b58e61"
ORCHARD_ZSA_BASE_PROOF_SIZE + ORCHARD_ZSA_PER_ACTION_PROOF_SIZE * num_actions
```

The stale transaction-level `proofsOrchard` field inherited from the
base class is removed for V6, since V6 stores proofs inside the action
group.

This also regenerates the affected test vectors:

* `orchard_zsa_digests`
* `zip_0233`

in their supported output formats (Rust, JSON, and zcash).
